### PR TITLE
Feature/course entity

### PR DESCRIPTION
**Pull request by nugentmichael1**
Check the application.properties file for local database credentials.

### CODE REVIEW

File: src/main/java/com/revlearn/team1/controller/CourseController.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import com.revlearn.team1.model.Course;
+import com.revlearn.team1.service.CourseService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/Course")
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/controller/CourseController.java:12]: can you lowercase this request_mapping to /courses

File: src/main/java/com/revlearn/team1/controller/CourseController.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+@RequiredArgsConstructor
+public class CourseController {
+    private final CourseService courseService;
+
+    @GetMapping("/{id}")
+    public Course getCourseById(@RequestParam Long courseId) {
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/controller/CourseController.java:18]: can you replace @RequestParam with @PathVariable

File: src/main/java/com/revlearn/team1/controller/CourseController.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        return courseService.getById(courseId);
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/controller/CourseController.java:19]: can you A) wrap the return in 'ResponseEntity' to handle HTTP status codes
and
B) add a null check and return a 404 if course doesn't exist

File: src/main/java/com/revlearn/team1/exceptions/CourseNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.revlearn.team1.exceptions;
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/exceptions/CourseNotFoundException.java:1]: OOH I love the custom exception handling!! nice touch

File: src/main/java/com/revlearn/team1/controller/TransactionController.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.revlearn.team1.controller;
+
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/v1/transaction")
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/controller/TransactionController.java:7]: I know this isnt your controller and its just so you don't get compile errors, but can we keep the RequestMapping convention consistent? either change /courses to /api/courses or api/transactions to /transactions

File: src/main/java/com/revlearn/team1/exceptions/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.revlearn.team1.exceptions;
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/exceptions/GlobalExceptionHandler.java:1]: love it!!

File: src/main/java/com/revlearn/team1/service/CourseService.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+import com.revlearn.team1.exceptions.CourseNotFoundException;
+import com.revlearn.team1.model.Course;
+import com.revlearn.team1.repository.CourseRepo;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CourseService {
+    CourseRepo courseRepo;
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/service/CourseService.java:12]: my linter is telling me: CourseRepo should be 'final' to ensure it is injected via constructor and can't be changed

File: src/main/java/com/revlearn/team1/service/TransactionService.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.revlearn.team1.service;
+
+import org.springframework.stereotype.Service;
+
+@Service
+public interface TransactionService
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/service/TransactionService.java:6]: nice job with the interface! that's gonna make testing hella easier!

File: src/main/java/com/revlearn/team1/model/Course.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+@Data
+public class Course {
+
+    @ManyToMany
+    @JoinTable(
+            name = "course_educators",
+            joinColumns = @JoinColumn(name = "course_id"),
+            inverseJoinColumns = @JoinColumn(name = "educator_id")
+    )
+    private final Set<User> educators = new HashSet<>();
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/model/Course.java:24]: my linter is telling me you need to remove the final fields here: 'if you need to use Lombok’s @Data, remove final from the fields. final fields cannot be modified after initialization, which is incompatible with the typical setter behavior.'

Comment (treycrossley) [reply src/main/java/com/revlearn/team1/model/Course.java:24]: speicifically from educators, students, DiscussionPosts

File: src/main/java/com/revlearn/team1/model/TransactionModel.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+@Entity
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@ToString
+@EqualsAndHashCode
+public class TransactionModel
+{
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
```
Comment (treycrossley) [inline src/main/java/com/revlearn/team1/model/TransactionModel.java:16]: every other id used GenerationType.Identity, so lets keep it consistent, yeah?

**Review by treycrossley — CHANGES_REQUESTED** [review]

File: src/main/java/com/revlearn/team1/controller/CourseController.java (hunk 3 above)
Comment (nugentmichael1) [reply src/main/java/com/revlearn/team1/controller/CourseController.java:19]: The HTTP status codes are handled by the exception handler in the Global Exception Handler.  Any errors will result in HTTP Response Entities in that file.

Comment (treycrossley) [reply src/main/java/com/revlearn/team1/controller/CourseController.java:19]: oh ur right! Dope! ignore this comment then

File: src/main/java/com/revlearn/team1/service/TransactionService.java (hunk 8 above)
Comment (nugentmichael1) [reply src/main/java/com/revlearn/team1/service/TransactionService.java:6]: This is Daniel's work.  I pulled his branch into mine.

Comment (treycrossley) [reply src/main/java/com/revlearn/team1/service/TransactionService.java:6]: TAKE THE DAMN CREDIT

**Review by nugentmichael1 — COMMENTED** [review]
Good notes. I implemented all corrections except for the HTTP failure codes through exception handlers.

**Review by treycrossley — APPROVED** [review]

**Comment by treycrossley** [comment]
Looks great! :) also love the DTO addition

**Pull request merged**